### ISSUE TITLE
Notices when using Router Generator with multiple middlewares

### ISSUE THREAD

**Issue by prolic** [zendframework/zend-expressive-aurarouter#8]
We have a route like this:

```
[
    'name' => 'page::user-todo-list',
    'path' => '/user-todo-list/{user_id}',
    'middleware' => \Prooph\ProophessorDo\App\Action\UserTodoList::class,
    'allowed_methods' => ['GET'],
    'options' => [
        'tokens' => [
            'user_id' => '[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}'
        ],
    ],
],
```

And this produces notices:

```
PHP Notice:  Array to string conversion in /var/www/proophessor-do/vendor/aura/router/src/Generator.php on line 131
```

This can cause trouble as routes are assembled and injected into JS code so the notice breaks JS syntax then.

When using multiple middlewares like:

```
'middleware' => [
    \Prooph\ProophessorDo\Middleware\JsonPayload::class,
    \Prooph\Psr7Middleware\CommandMiddleware::class,
],
```

the router gets an array of {action} params and will create this notice.

see also: https://github.com/prooph/proophessor-do/issues/64

**Comment by sandrokeil** [comment]
Confirmed. This is only caused if you call `\Zend\Expressive\Router\RouterInterface::generateUri()` with AuraRouter, but not with the url view helper. The array middleware execution works fine.

Example with Zend Expressive Skeleton

``` php
    'routes' => [
        [
            'name' => 'home',
            'path' => '/',
            'middleware' => [
                App\Action\HomePageAction::class // <-- this array causes the error
            ],
            'allowed_methods' => ['GET'],
        ],
    ],
```

``` php
// error
$container->get(RouterInterface::class)->generateUri('home');

// no error
$container->get(UrlHelper::class)->__invoke('home');
```

**Comment by sandrokeil** [comment]
The `UrlHelper` provides `$params` with `array(2) { ["action"]=> array(1) { [0]=> string(25) "App\Action\HomePageAction" } ["REQUEST_METHOD"]=> string(3) "GET" }`. Seems that this is the problem, if we use the Router and omit  `$params`.

**Comment by harikt** [comment]
Yesterday I looked into the issue.

@sandrokeil is correct.

Internally the generator was trying to convert some of the values https://github.com/auraphp/Aura.Router/blob/d8cb21b7fa8d2b2ae874093d5440688b3356f9c4/src/Generator.php#L131 

``` php
<?php
$k = [
    '{action}' => [
        'Action',
    ]
];

echo strtr('/', $k);
```

In our case we are passing an array to `{action}` for it is an array of middleware. 

The `strtr` throws warnings when your array is multidimensional . First I thought of not passing the action to the router add method. But that seems not good for expressive will try to get the middleware from the RouteResult. Another fix is check if `{action}` is an array and unset it before calling `strtr` .

And this is my final impression. This is using 2.x of Aura.Router and considering https://github.com/zendframework/zend-expressive-aurarouter/pull/7 still not fixed due to BC breaks probably it is not worth to fix for 2.x . I would recommend pushing 3.x (  https://github.com/zendframework/zend-expressive-aurarouter/pull/11 ) of Aura.Router which will be maintained.

**Comment by weierophinney** [comment]
This should be resolved with #11.
